common/network-common/src/main/java/org/apache/spark/network/sasl/aes/AesEncryption.java

```diff
@@ -0,0 +1,176 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.network.sasl.aes;
+
+import java.io.IOException;
+import java.nio.channels.WritableByteChannel;
+
+import com.google.common.base.Preconditions;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.*;
+import io.netty.util.AbstractReferenceCounted;
+import org.apache.commons.crypto.stream.CryptoInputStream;
+import org.apache.commons.crypto.stream.CryptoOutputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.spark.network.util.ByteArrayReadableChannel;
+import org.apache.spark.network.util.ByteArrayWritableChannel;
+
+public class AesEncryption {
+  public static final String ENCRYPTION_HANDLER_NAME = "AesEncryption";
+  public static final String DECRYPTION_HANDLER_NAME = "AesDecryption";
+
+  public static void addToChannel(Channel ch, AesCipher cipher) {
+    ch.pipeline().addFirst(ENCRYPTION_HANDLER_NAME, new AesEncryptHandler(cipher))
+      .addFirst(DECRYPTION_HANDLER_NAME, new AesDecryptHandler(cipher));
+  }
+
+  private static class AesEncryptHandler extends ChannelOutboundHandlerAdapter {
+    private final AesCipher cipher;
+    private ByteArrayWritableChannel byteChannel;
+
+    AesEncryptHandler(AesCipher cipher) {
+      this.cipher = cipher;
+      byteChannel = new ByteArrayWritableChannel(AesCipher.STREAM_BUFFER_SIZE);
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
+      throws Exception {
+      ctx.write(new EncryptMessage(cipher, msg, byteChannel), promise);
+    }
+  }
+
+  private static class AesDecryptHandler extends ChannelInboundHandlerAdapter {
+    private final AesCipher cipher;
+    private CryptoInputStream cis;
+    private ByteArrayReadableChannel byteChannel;
+
+    AesDecryptHandler(AesCipher cipher) {
+      this.cipher = cipher;
+      byteChannel = new ByteArrayReadableChannel();
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object data) throws Exception {
+      ByteBuf in = (ByteBuf) data;
+
+      try {
+        while (in.isReadable()) {
+          byteChannel.feedData(in);
+          cis = cipher.CreateInputStream(byteChannel);
```

Comment:
Why isn't this done once in the constructor? Then you can explicitly close it in the `channelInactive` callback, and avoid the cache in `AesCipher` too.